File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/model/JcrCrudManager.java
package solutions.digamma.damas.jcr.model;

import solutions.digamma.damas.entity.CrudManager;
import solutions.digamma.damas.common.WorkspaceException;
import solutions.digamma.damas.entity.Entity;
import solutions.digamma.damas.auth.Token;
import solutions.digamma.damas.inspection.NotNull;
import solutions.digamma.damas.jcr.common.Exceptions;
import solutions.digamma.damas.jcr.session.SessionWrapper;
import solutions.digamma.damas.logging.Logged;

import javax.jcr.RepositoryException;
import javax.jcr.Session;

/**
 * JCR CRUD manager.
 *
 * @author Ahmad Shahwan
 */
abstract public class JcrCrudManager<T extends Entity>
        extends JcrReadManager<T> implements CrudManager<T> {

    @Logged
    @Override
    public T create(@NotNull Token token, @NotNull T entity)
            throws WorkspaceException {
        try (SessionWrapper session = openSession(token)) {
            return this.create(session.getSession(), entity);
        } catch (RepositoryException e) {
            throw Exceptions.convert(e);
        }
    }

    @Logged
    @Override
    public T update(@NotNull Token token, @NotNull String id, @NotNull T entity)
            throws WorkspaceException {
        try (SessionWrapper session = openSession(token)) {
            return this.update(session.getSession(), id, entity);
        } catch (RepositoryException e) {
            throw Exceptions.convert(e);
        }
    }

    @Logged
    @Override
    public void delete(@NotNull Token token, @NotNull String id)
            throws WorkspaceException {
        try (SessionWrapper session = openSession(token)) {
            this.delete(session.getSession(), id);
        } catch (RepositoryException e) {
            throw Exceptions.convert(e);
        }
    }

    /**
     * Perform creation.
     *
     * @param session
     * @param entity
     * @return
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    abstract protected T create(
            @NotNull Session session, @NotNull T entity)
            throws RepositoryException, WorkspaceException;

    /**
     * Perform update.
     *
     * @param session
     * @param id
     * @param entity
     * @return
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    abstract protected T update(
            @NotNull Session session,
            @NotNull String id,
            @NotNull T entity)
            throws RepositoryException, WorkspaceException;

    /**
     * Perform deletion.
     *
     * @param session
     * @param id
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    abstract protected void delete(
            @NotNull Session session, @NotNull String id)
            throws RepositoryException, WorkspaceException;
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/model/JcrCrudManager.kt
package solutions.digamma.damas.jcr.model

import solutions.digamma.damas.common.WorkspaceException
import solutions.digamma.damas.entity.CrudManager
import solutions.digamma.damas.entity.Entity
import solutions.digamma.damas.jcr.common.Exceptions
import solutions.digamma.damas.logging.Logged
import javax.jcr.RepositoryException

/**
 * JCR CRUD manager.
 *
 * @author Ahmad Shahwan
 */
internal abstract class JcrCrudManager<T: Entity>: JcrReadManager<T>(), CrudManager<T> {

    @Logged
    @Throws(WorkspaceException::class)
    override fun create(entity: T): T = Exceptions.check {
        this.doCreate(entity)
    }


    @Logged
    @Throws(WorkspaceException::class)
    override fun update(id: String, entity: T): T = Exceptions.check {
        this.doUpdate(id, entity)
    }

    @Logged
    @Throws(WorkspaceException::class)
    override fun delete(id: String): Unit = Exceptions.check {
        this.doDelete(id)
    }

    /**
     * Perform creation.
     *
     * @param pattern
     * @return
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    @Throws(RepositoryException::class, WorkspaceException::class)
    protected abstract fun doCreate(pattern: T): T

    /**
     * Perform update.
     *
     * @param id
     * @param pattern
     * @return
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    @Throws(RepositoryException::class, WorkspaceException::class)
    protected abstract fun doUpdate(id: String, pattern: T): T

    /**
     * Perform deletion.
     *
     * @param id
     * @throws RepositoryException
     * @throws WorkspaceException
     */
    @Throws(RepositoryException::class, WorkspaceException::class)
    protected abstract fun doDelete(id: String)
}
